File: day04/js/solution.js
import fs from "fs";

const grid = fs
  .readFileSync("../input.txt", "utf-8")
  .trim()
  .split("\n");

const directions = [
  [-1, -1], [-1, 0], [-1, 1],
  [ 0, -1],          [ 0, 1],
  [ 1, -1], [ 1, 0], [ 1, 1],
];

function countNeighbors(grid, r, c) {
  let count = 0;
  for (const [dr, dc] of directions) {
    const nr = r + dr;
    const nc = c + dc;
    if (
      nr >= 0 && nr < grid.length &&
      nc >= 0 && nc < grid[0].length &&
      grid[nr][nc] === "@"
    ) {
      count++;
    }
  }
  return count;
}

function part1(grid) {
  let result = 0;

  for (let r = 0; r < grid.length; r++) {
    for (let c = 0; c < grid[0].length; c++) {
      if (grid[r][c] === "@" && countNeighbors(grid, r, c) < 4) {
        result++;
      }
    }
  }

  return result;
}

function part2(originalGrid) {
  const grid = originalGrid.map(row => [...row]);
  let totalRemoved = 0;

  while (true) {
    const toRemove = [];

    for (let r = 0; r < grid.length; r++) {
      for (let c = 0; c < grid[0].length; c++) {
        if (grid[r][c] === "@" && countNeighbors(grid, r, c) < 4) {
          toRemove.push([r, c]);
        }
      }
    }

    if (toRemove.length === 0) break;

    for (const [r, c] of toRemove) {
      grid[r][c] = ".";
    }

    totalRemoved += toRemove.length;
  }

  return totalRemoved;
}

console.log("Part 1:", part1(grid));
console.log("Part 2:", part2(grid));

File: day04/py/solution.py
from pathlib import Path

grid = [list(line) for line in Path("../input.txt").read_text().splitlines()]

DIRECTIONS = [
    (-1, -1), (-1, 0), (-1, 1),
    ( 0, -1),          ( 0, 1),
    ( 1, -1), ( 1, 0), ( 1, 1),
]

def count_neighbors(grid, r, c):
    rows, cols = len(grid), len(grid[0])
    count = 0
    for dr, dc in DIRECTIONS:
        nr, nc = r + dr, c + dc
        if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == "@":
            count += 1
    return count

def part1(grid):
    result = 0
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == "@" and count_neighbors(grid, r, c) < 4:
                result += 1
    return result

def part2(original_grid):
    grid = [row[:] for row in original_grid]
    total_removed = 0

    while True:
        to_remove = []

        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == "@" and count_neighbors(grid, r, c) < 4:
                    to_remove.append((r, c))

        if not to_remove:
            break

        for r, c in to_remove:
            grid[r][c] = "."

        total_removed += len(to_remove)

    return total_removed

print("Part 1:", part1(grid))
print("Part 2:", part2(grid))
